### PR TITLE
Added perplexity eval

### PR DESCRIPTION
**Pull request by pgarbacki**

### CODE REVIEW

**Review by benjibc — APPROVED** [review]

File: recipes/eval/perplexity/conf/model/codellama-7b.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  - ../../../../common/conf/model/codellama-7b@_here_
+
+name: codellama-7b
+
+flash_attention: True
+torch_dtype: bfloat16
+
+rope_scaling:
+  type: linear
+  factor: 2.0
```
Comment (dzhulgakov) [inline recipes/eval/perplexity/conf/model/codellama-7b.yaml:11]: so you do apply scaling, don't you?

File: recipes/eval/perplexity/eval.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    """
+    stride = config.stride
+    num_tokens = config.num_tokens
+    perplexity = collections.defaultdict(int)
+    device = torch.cuda.current_device()
+    total_input_tokens = 0
+    for text in tqdm(texts, desc="computing perplexity"):
+        input_ids = tokenizer(text, return_tensors="pt")["input_ids"]
+        num_input_tokens = input_ids.size(1)
+        total_input_tokens += num_input_tokens
```
Comment (dzhulgakov) [inline recipes/eval/perplexity/eval.py:88]: not used?

File: recipes/eval/perplexity/eval.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        input_ids = input_ids[:, :num_tokens].to(device)
+        torch.cuda.reset_peak_memory_stats()
+        for num_tokens in range(stride, num_tokens + 1, stride):
+            local_input_ids = input_ids[:, :num_tokens]
+            with torch.no_grad():
+                loss = model(input_ids=local_input_ids, labels=local_input_ids).loss
+            perplexity[num_tokens] += torch.exp(loss).item()
+    perplexity = {k: v / len(texts) for k, v in perplexity.items()}
+    return {
+        "avg_input_tokens": num_input_tokens / len(texts),
```
Comment (dzhulgakov) [inline recipes/eval/perplexity/eval.py:102]: isn't it last text's length?

File: recipes/eval/perplexity/conf/model/codellama-7b.yaml (hunk 1 above)
Comment (pgarbacki) [reply recipes/eval/perplexity/conf/model/codellama-7b.yaml]: I tried with and without scaling.

File: recipes/eval/perplexity/eval.py (hunk 2 above)
Comment (pgarbacki) [reply recipes/eval/perplexity/eval.py:88]: fixed.

File: recipes/eval/perplexity/eval.py (hunk 3 above)
Comment (pgarbacki) [reply recipes/eval/perplexity/eval.py]: fixed

File: recipes/eval/perplexity/eval.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        input_ids = tokenizer(text, return_tensors="pt")["input_ids"]
+        num_input_tokens = input_ids.size(1)
+        total_input_tokens += num_input_tokens
+        if num_input_tokens < num_tokens:
+            raise ValueError(
+                f"expected at least {num_tokens}, found {num_input_tokens}"
+            )
+        input_ids = input_ids[:, :num_tokens].to(device)
+        torch.cuda.reset_peak_memory_stats()
+        for num_tokens in range(stride, num_tokens + 1, stride):
```
Comment (dzhulgakov) [inline recipes/eval/perplexity/eval.py:95]: you override `num_tokens` here. It kind of works out in practice, but still bad code

**Pull request merged**